### PR TITLE
sql: add batches read to {join,table}reader

### PR DESCRIPTION
**Pull request by jordanlewis**
Based on #54847

... and fix a small error in bytes read.

This commit adds bytes read to the join reader, so it's possible to see
how many index bytes were read from the looked-up index. It also fixes a
small bug in bytes read: previously, if a scan had been aborted half way
through (due to e.g. hitting a limit) the number of bytes read could
have been less than what was actually read from disk, due to buffering
in the kv batch fetcher.

Also, add a "batches read" field to the tablereader and joinreader, so
it's possible to see how many "batches" (round trips to kv) were done in
each operator. This makes it easier to see how much transient memory was
buffered in the readers.

Release note (sql change): add "batches read" to EXPLAIN ANALYZE and
traces, and add "bytes read" to the join reader from which it was
missing.

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/row/kv_batch_fetcher.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKmIHHAeyY9PI3r_Id:-MIKmIHHAeyY9PI3r_Ie:brm2ofw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/row/kv_batch_fetcher.go#L80)):*
> ```Go
> 	acc              mon.BoundAccount
> 	// bytesRead is the number of bytes read from disk of the responses that have
> 	// been returned to the client so far. Note that, to get an accurate tally of
> ```

~nit: this comment seems to contradict itself a bit, i.e. "that have been returned to the client ... responses that haven't yet been returned". I'd reword it.~ Oh, I see the meaning, I find the comment a bit confusing (the second sentence), maybe something like this would be better: `Note that this *might* be different from the total number of bytes read.` and I'd move the note about iteration to `getBytesRead` method.
___
*[pkg/sql/rowexec/joinreader.go, line 589 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKo7Hs8bAzuYQBeEbe:-MIKo7Hs8bAzuYQBeEbf:b-e3vl86) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/joinreader.go#L589)):*
> ```Go
> func (jrs *JoinReaderStats) Stats() map[string]string {
> 	statsMap := jrs.InputStats.Stats(joinReaderTagPrefix)
> 	toMerge := jrs.IndexLookupStats.IndexLookupStats.Stats(joinReaderIndexTagPrefix)
> ```

Why do we have repeated `IndexLookupStats`?
___
*[pkg/sql/rowexec/stats.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoHHA8eOs5GddW_hS:-MIKoHHA8eOs5GddW_hT:b-8llkml) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.go#L106)):*
> ```Go
> 	bytesReadQueryPlanSuffix = "bytes read"
> 	// batchesReadQueryPlanSuffix is the tag suffix for the bytes read.
> 	batchesReadQueryPlanSuffix = "batches read"
> ```

I think this might be confusing with the concept of "batches" in the vectorized engine (and on the diagrams), maybe `s/batches read/KV batches read/`?
___
*[pkg/sql/rowexec/stats.proto, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoc8bD-L1N3uUmtVT:-MIKoc8bD-L1N3uUmtVU:b-bqqpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.proto#L39)):*
> ```protobuf
>   reserved 2;
>   reserved 3;
>   JoinReaderIndexStats index_lookup_stats = 4 [(gogoproto.nullable) = false];
> ```

I see, maybe just embed the contents of `JoinReaderIndexStats` into `JoinReaderStats`?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/row/kv_batch_fetcher.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKmIHHAeyY9PI3r_Id:-MIMNW6KDPy01Hav1ouh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/row/kv_batch_fetcher.go#L80)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

~nit: this comment seems to contradict itself a bit, i.e. "that have been returned to the client ... responses that haven't yet been returned". I'd reword it.~ Oh, I see the meaning, I find the comment a bit confusing (the second sentence), maybe something like this would be better: `Note that this *might* be different from the total number of bytes read.` and I'd move the note about iteration to `getBytesRead` method.
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 589 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKo7Hs8bAzuYQBeEbe:-MIMNWd4AZlId-bsLiFe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/joinreader.go#L589)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Why do we have repeated `IndexLookupStats`?
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/stats.go, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoHHA8eOs5GddW_hS:-MIMNWyq6LBhUAejQuvJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.go#L106)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this might be confusing with the concept of "batches" in the vectorized engine (and on the diagrams), maybe `s/batches read/KV batches read/`?
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/stats.proto, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoc8bD-L1N3uUmtVT:-MIMNXV_AKfjVB6_Ay-t:bepcck5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.proto#L39)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I see, maybe just embed the contents of `JoinReaderIndexStats` into `JoinReaderStats`?
</blockquote></details>

Can't embed in protos as far as I can tell, I renamed to `stats`.

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 9 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/rowexec/stats.proto, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoc8bD-L1N3uUmtVT:-MIM_5Dl8RQ0uRMi-31v:bwqnzkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.proto#L39)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can't embed in protos as far as I can tell, I renamed to `stats`.
</blockquote></details>

I meant not having a separate message but "embedding" the contents, i.e.:
```
message JoinReaderStats {
  InputStats input_stats ...
  reserved 2,3;
  InputStats index_lookup_stats = 4 ...
  int64 bytes_read = 5;
  int64 batches_read = 6;
}
```
I just don't see what benefit we're getting from defining a separate protobuf.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/rowexec/stats.proto, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-MIKoc8bD-L1N3uUmtVT:-MIMspEc-GTAozCJe0sg:bfeudqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a7831b5aed610189f5c83d2af72a420a1dd9958/pkg/sql/rowexec/stats.proto#L39)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I meant not having a separate message but "embedding" the contents, i.e.:
```
message JoinReaderStats {
  InputStats input_stats ...
  reserved 2,3;
  InputStats index_lookup_stats = 4 ...
  int64 bytes_read = 5;
  int64 batches_read = 6;
}
```
I just don't see what benefit we're getting from defining a separate protobuf.
</blockquote></details>

Oh I see what you mean, that's a good point. Fixed.

**Review by yuzefovich — COMMENTED** [review]
Last couple of questions:
- I think now it is possible that "bytes read" information is duplicated when we have a vectorized flow and wrapped `tableReader` or `joinReader`, could you please check that case? I feel like we need to update logic in the vectorized case.
- on a related note, it'd be nice to extend this PR to show number of "kv batches read" to the vectorized flows. I think that those should show up for free on wrapped `rowexec.tableReader`s and join readers, but we need some more plumbing to support `ColBatchScan`.

> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54848#-:-MIMxB-D7bduwzicvPI-:b-hsy3vi)* status: :shipit: complete! 0 of 0 LGTMs obtained
